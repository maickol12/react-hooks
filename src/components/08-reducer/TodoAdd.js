import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo}) => {
    const [values, handleInputChange, reset] = useForm({ description: '' });

    const handleSubmit = (e) =>{
        e.preventDefault();
        
        if(values.description.trim().length <= 1){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: values.description,
            done: false
        };
    

        handleAddTodo(newTodo);
        reset();
    }
    return (
        <>
           <h4>Agregar todo</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input
                    className='form-control'
                    type='text'
                    value={ values.description }
                    onChange={ handleInputChange }
                    name='description'
                    placeholder='Aprender...'
                    autoComplete='off'></input>
                <button type='submit' className='btn btn-primary mt-1 btn-block'>
                    Agregar
                </button>
            </form>  
        </>
    )
}
