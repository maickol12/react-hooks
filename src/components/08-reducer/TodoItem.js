import React from 'react'

export const TodoItem = ({value,handleToggle,deleteTodo,indice}) => {
    const { id, done, desc } = value;
    return (
        <li key={id} className='list-group-item'>
            <p 
                className={`${done && 'complete'}`}
                onClick={() => handleToggle(id) }>
                {indice + 1}. {desc}
            </p>
            <button 
                onClick={ () => deleteTodo(id) }
                className='btn btn-danger'>
                Borrar
            </button>
        </li>
        
    )
}
