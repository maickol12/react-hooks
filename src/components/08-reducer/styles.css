body{
    padding: 70px;
}

p{
    margin: 0px;
    cursor: pointer;
}

.complete{
    text-decoration: line-through;
}

.list-group-item{
    align-items: center;
    display: flex;
    justify-content: space-between;
}