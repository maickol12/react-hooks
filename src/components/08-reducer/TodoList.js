import React from 'react';
import { TodoItem } from './TodoItem';

export const TodoList = ({todos,handleToggle,deleteTodo}) => {
    return (
        <div className='col-5'>
            <ul className='list-group list-group-flush'>
            { todos.map((value,indice) => (
                       <TodoItem
                            key={`todoItem${value.id}`}
                            value={value}
                            handleToggle={handleToggle}
                            deleteTodo={deleteTodo}
                            indice={indice} /> 
                    )
                    ) 
                }
            </ul>
        </div>
    )
}
