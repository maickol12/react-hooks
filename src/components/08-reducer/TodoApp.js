import React, { useReducer, useState, useEffect } from 'react'
import { useForm } from '../../hooks/useForm';
import './styles.css';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';


const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || []
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'Aprendiendo react',
    //     done: false
    // }];
}


export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos',JSON.stringify(todos));
    }, [todos])



    const deleteTodo = (id) => {
        
        const action = {
            type:'delete',
            payload: id
        }

        dispatch(action);

    }

    const handleToggle = (id) => {
        dispatch({type:'toggle',payload:id});
    }
    
    const handleAddTodo = (newTodo) => {
        const action = {
            type:'add',
            payload: newTodo
        }
        dispatch(action);
    }

    return (
        <div>
            <h1>TodoAPP ({ todos.length})</h1>
            <hr />
            <div className='row'>
                <TodoList 
                    todos={todos}
                    handleToggle={handleToggle}
                    deleteTodo={deleteTodo} />
                <div className='col-7'>
                   <TodoAdd handleAddTodo={handleAddTodo} />
                </div>

            </div>
        </div>
    )
}
