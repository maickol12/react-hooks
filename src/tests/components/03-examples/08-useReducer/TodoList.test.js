import { shallow } from "enzyme"
import { TodoList } from "../../../../components/08-reducer/TodoList"
import { demosTodos } from "../../../fixtures/demosTodos"

describe('Pruebas en <TodoList />', () => {
    const handleToggle  = jest.fn();
        const deleteTodo    = jest.fn();
        const wrapper = shallow( <TodoList 
            todos = {demosTodos} 
            handleToggle={handleToggle}
            deleteTodo={deleteTodo}
             /> );
    test('Debe de mostrarse correctamente', () => {
    
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe de tener dos <TodoList />', () => {
        expect(wrapper.find('TodoItem').length).toBe(demosTodos.length);
        expect(wrapper.find('TodoItem').at(0).prop('handleDelete')).toEqual(expect.any(Function));
    });
    
})
