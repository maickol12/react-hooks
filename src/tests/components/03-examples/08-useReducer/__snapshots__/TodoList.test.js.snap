// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoList /> Debe de mostrarse correctamente 1`] = `
<div
  className="col-5"
>
  <ul
    className="list-group list-group-flush"
  >
    <TodoItem
      deleteTodo={[MockFunction]}
      handleToggle={[MockFunction]}
      indice={0}
      key="todoItem1"
      value={
        Object {
          "desc": "Aprender React",
          "done": false,
          "id": 1,
        }
      }
    />
    <TodoItem
      deleteTodo={[MockFunction]}
      handleToggle={[MockFunction]}
      indice={1}
      key="todoItem2"
      value={
        Object {
          "desc": "Aprender Mongo",
          "done": false,
          "id": 2,
        }
      }
    />
  </ul>
</div>
`;

exports[`Pruebas en <TodoList /> Debe de tener dos <TodoList /> 1`] = `
<div
  className="col-5"
>
  <ul
    className="list-group list-group-flush"
  >
    <TodoItem
      deleteTodo={[MockFunction]}
      handleToggle={[MockFunction]}
      indice={0}
      key="todoItem1"
      value={
        Object {
          "desc": "Aprender React",
          "done": false,
          "id": 1,
        }
      }
    />
    <TodoItem
      deleteTodo={[MockFunction]}
      handleToggle={[MockFunction]}
      indice={1}
      key="todoItem2"
      value={
        Object {
          "desc": "Aprender Mongo",
          "done": false,
          "id": 2,
        }
      }
    />
  </ul>
</div>
`;
